Source/WebCore/layout/formattingContexts/inline/InlineLine.cpp

```diff
@@ -297,6 +298,16 @@ void Line::appendInlineBoxStart(const InlineItem& inlineItem, const RenderStyle&
     // This is really just a placeholder to mark the start of the inline box <span>.
     ++m_nonSpanningInlineLevelBoxCount;
     auto logicalLeft = lastRunLogicalRight();
+    // https://www.w3.org/TR/css-text-4/#text-autospace-property
+    // The InlineBoxStart's text spacing should contribute to its logical left becase we need to represent a spacing at the element boundary between this InlineItemStartBox's element and the previous element. However, we will trim this spacing if this InlineBoxStart is the first box on the line since text-spacing only affects adjacent characters on the same line.
+    const auto&  inlineItems  = formattingContext().inlineContentCache().inlineItems();
+    if (!inlineItems.elementBoundaryTextSpacings().isEmpty() && !m_runs.isEmpty()) {
+        unsigned indexOnInlineItemList = std::distance(inlineItems.content().begin(), &inlineItem);
+        const auto& elementBoundarySpacings = inlineItems.elementBoundaryTextSpacings();
+        if (auto elementBoundarySpacing = elementBoundarySpacings.find(indexOnInlineItemList); elementBoundarySpacing !=  elementBoundarySpacings.end())
+            logicalLeft += elementBoundarySpacing->value;
+    }
+
```

Comment:
since this will certainly not work with run reordering, it might be better to not introduce this level of complexity for the non-bidi case only and represent the gap as logical width instead. -which means Line does not need to know about this at all (as we discussed offline, Line is just a representation of widths in logical order with no real horizontal placement).